File: src/services/ai-logs.ts
import { ipc } from "@/lib/ipc";

// ── Types ──────────────────────────────────────────────────

export interface LlmCall {
  id: string;
  session_id: string;
  agent_role: string;
  model: string;
  provider: string;
  system_prompt: string | null;
  messages_json: string;
  tools_json: string | null;
  temperature: number | null;
  max_tokens: number | null;
  response_content: string | null;
  response_tool_calls: string | null;
  finish_reason: string | null;
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  started_at: string;
  completed_at: string | null;
  latency_ms: number | null;
  status: string;
  error_message: string | null;
  metadata: string;
  created_at: string;
}

export interface ToolExecution {
  id: string;
  session_id: string;
  llm_call_id: string | null;
  tool_name: string;
  arguments_json: string;
  result_content: string | null;
  is_error: boolean;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
  duration_ms: number | null;
  sandbox_allowed: boolean;
  sandbox_violation: string | null;
  pve_blocked: boolean;
  metadata: string;
  created_at: string;
}

export interface TokenUsageStats {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  tools: {
    total_calls: number;
    errors: number;
    sandbox_blocked: number;
    success_rate: number;
  };
  models: Array<{
    model: string;
    calls: number;
    input_tokens: number;
    output_tokens: number;
    avg_latency_ms: number | null;
  }>;
}

export interface SandboxViolation {
  id: string;
  session_id: string | null;
  violation_type: string;
  resource: string;
  action: string;
  rule_matched: string | null;
  tool_name: string | null;
  arguments_json: string | null;
  detected_at: string;
  created_at: string;
}

// ── API Functions ──────────────────────────────────────────

export async function getLlmCalls(sessionId: string, limit?: number): Promise<LlmCall[]> {
  return ipc<LlmCall[]>("ai_log_llm_calls", { sessionId, limit: limit ?? 50 });
}

export async function getToolExecutions(sessionId: string, limit?: number): Promise<ToolExecution[]> {
  return ipc<ToolExecution[]>("ai_log_tool_executions", { sessionId, limit: limit ?? 50 });
}

export async function getTokenUsage(sessionId: string): Promise<TokenUsageStats> {
  return ipc<TokenUsageStats>("ai_log_token_usage", { sessionId });
}

export async function getSandboxViolations(sessionId: string, limit?: number): Promise<SandboxViolation[]> {
  return ipc<SandboxViolation[]>("ai_log_sandbox_violations", { sessionId, limit: limit ?? 50 });
}

File: src-tauri/src/app/commands/ai_logs.rs
use crate::errors::{AppError, IpcResponse};
use crate::AppState;
use tauri::State;

#[tauri::command]
pub async fn ai_log_llm_calls(
    state: State<'_, AppState>,
    session_id: String,
    limit: Option<u32>,
) -> Result<IpcResponse<Vec<crate::infra::db::ai_log_store::LlmCall>>, AppError> {
    let db = state.db.lock().await;
    let calls = db.get_llm_calls(&session_id, limit.unwrap_or(50))?;
    Ok(IpcResponse::ok(calls))
}

#[tauri::command]
pub async fn ai_log_tool_executions(
    state: State<'_, AppState>,
    session_id: String,
    limit: Option<u32>,
) -> Result<IpcResponse<Vec<crate::infra::db::ai_log_store::ToolExecution>>, AppError> {
    let db = state.db.lock().await;
    let execs = db.get_tool_executions(&session_id, limit.unwrap_or(50))?;
    Ok(IpcResponse::ok(execs))
}

#[tauri::command]
pub async fn ai_log_token_usage(
    state: State<'_, AppState>,
    session_id: String,
) -> Result<IpcResponse<serde_json::Value>, AppError> {
    let db = state.db.lock().await;
    let (input, output) = db.get_session_token_usage(&session_id)?;
    let tool_stats = db.get_session_tool_stats(&session_id)?;
    let model_stats = db.get_model_usage_stats(&session_id)?;
    Ok(IpcResponse::ok(serde_json::json!({
        "input_tokens": input,
        "output_tokens": output,
        "total_tokens": input + output,
        "tools": tool_stats,
        "models": model_stats,
    })))
}

#[tauri::command]
pub async fn ai_log_sandbox_violations(
    state: State<'_, AppState>,
    session_id: String,
    limit: Option<u32>,
) -> Result<IpcResponse<Vec<crate::infra::db::ai_log_store::SandboxViolation>>, AppError> {
    let db = state.db.lock().await;
    let mut stmt = db.conn.prepare(
        "SELECT id, session_id, violation_type, resource, action, rule_matched, tool_name, arguments_json, detected_at, created_at FROM sandbox_violations WHERE session_id = ?1 ORDER BY detected_at DESC LIMIT ?2"
    ).map_err(|e| AppError::internal(format!("Failed to prepare query: {}", e)))?;
    let rows = stmt.query_map(rusqlite::params![session_id, limit.unwrap_or(50)], |row| {
        Ok(crate::infra::db::ai_log_store::SandboxViolation {
            id: row.get(0)?,
            session_id: row.get(1)?,
            violation_type: row.get(2)?,
            resource: row.get(3)?,
            action: row.get(4)?,
            rule_matched: row.get(5)?,
            tool_name: row.get(6)?,
            arguments_json: row.get(7)?,
            detected_at: row.get(8)?,
            created_at: row.get(9)?,
        })
    }).map_err(|e| AppError::internal(format!("Failed to query violations: {}", e)))?;
    let violations = rows.collect::<Result<Vec<_>, _>>()
        .map_err(|e| AppError::internal(format!("Failed to collect violations: {}", e)))?;
    Ok(IpcResponse::ok(violations))
}
